drafts/2019-10-01-rewriting-irewritable.md

```diff
@@ -0,0 +1,448 @@
+---
+title: Rewriting IRewritable
+---
+
+> _This post is about my C# generic programming library [Sawmill](https://github.com/benjamin-hodgson/Sawmill). Have a read of [my earlier post](https://www.benjamin.pizza/posts/2017-11-13-recursion-without-recursion.html) for an introduction._
+
+## How things were
+
+I recently made a substantial change to Sawmill's core `IRewritable` interface. Here's what it used to look like:
+
+```csharp
+interface IRewritable<T> where T : IRewritable<T>
+{
+    Children<T> GetChildren();
+    T SetChildren(Children<T> newChildren);
+    T RewriteChildren(Func<T, T> transformer);
+}
+```
+
+It's a simple interface with two straightforward operations --- `GetChildren` returns the current object's immediate children and `SetChildren` replaces them. But there are also some warts.
+
+### Wart 1: `Children<T>`
+
+`Children<T>` is a custom struct, not a standard collection type:
+
+```csharp
+struct Children<T>
+{
+    public NumberOfChildren NumberOfChildren { get; }
+    public T First { get; }
+    public T Second { get; }
+    public ImmutableList<T> Many { get; }
+}
+enum NumberOfChildren
+{
+    None, One, Two, Many
+}
+```
+
+I wanted to avoid boxing when the object has a small number of children (which is fairly common in practice). So `GetChildren` returns a `Children<T>`, passing up to two children on the stack and the rest in an `ImmutableList`. The `NumberOfChildren` property tells the library how many of the struct's fields are filled in.
+
+This custom collection type is an extra hurdle to understanding `IRewritable`'s API. It also makes certain parts of Sawmill's implementation more complex --- many internal methods have to `switch` on the `NumberOfChildren` they're working with and do the same work in four three different ways. It's also relatively large for a `struct` (around four words) so there is a marginal performance cost associated with copying it around.
```

Comment:
"four three" -> "four"

Reply:
"four words" is an ambiguous, word size varies by architecture.  It's 3 pointers + 1 int, or 1 int + 2 * sizeof(T) + 1 pointer depending on what T is.  I'd say "at least 16 bytes*, often more" or similar.  (on a 32-bit config, each pointer will be 4 bytes + int which is always four).

As a random aside, it might actually have extra padding if naively laid out on 64-bit targets:  <4 bytes for NumberOfChildren> <4 bytes padding> <8 bytes First> <8 bytes Second> <8 bytes Many>.